test/lint/lint-filenames-permissions.py

```diff
@@ -0,0 +1,141 @@
+            )
+            exit_code = 1
+    return exit_code
+
+
+def check_all_file_permissions():
+    """
+    Checks all files in the repository match an allowed executable or non-executable file permission octal.
+
+    Additionally checks that for executable files, the file contains a shebang line
+    """
+    files = check_output(CMD_ALL_FILES, shell=True).decode("utf8").strip().split("\n")
+    exit_code = 0
+    for file in files:
+        file_extension = os.path.splitext(file)[1].strip(".")
+        file_perms = int(oct(os.stat(file).st_mode)[-3:])
+
+        if file_perms == ALLOWED_PERMISSION_EXECUTABLES:
+            shebang = open(file, "rb").readline().strip()
+
+            # For any file with executable permissions the first line must contain a shebang
+            if shebang[:2] != b"#!":
+                print(
+                    f"""File "{file}" has permission {ALLOWED_PERMISSION_EXECUTABLES} (executable) and is thus expected to contain a shebang '#!'. Add shebang or do "chmod {ALLOWED_PERMISSION_NON_EXECUTABLES} {file}" to make it non-executable."""
+                )
+                exit_code = 1
+
+            # For certain file extensions that have been defined, we also check that the shebang conforms to a specific
+            # allowable set of shebangs
+            if file_extension in ALLOWED_EXECUTABLE_SHEBANG.keys():
+                if shebang not in ALLOWED_EXECUTABLE_SHEBANG[file_extension]:
+                    print(
+                        f"""File "{file}" is missing expected shebang """
+                        + " or ".join(
+                            [x.decode("utf-8") for x in ALLOWED_EXECUTABLE_SHEBANG[file_extension]]
+                        )
+                    )
+
+        elif file_perms == ALLOWED_PERMISSION_NON_EXECUTABLES:
+            continue
+        else:
+            print(
+                f"""File '{file}' has unexpected permission {file_perms}. Do "chmod {ALLOWED_PERMISSION_NON_EXECUTABLES} {file}" (if non-executable) or "chmod {ALLOWED_PERMISSION_EXECUTABLES} {file}" (if executable)."""
+            )
+            exit_code = 1
+
+    return exit_code
+
+
+def check_shebang_file_permissions():
+    """
+    Checks every file that contains a shebang line to ensure it has an executable permission
+    """
+    files = check_output(CMD_SHEBANG_FILES, shell=True).decode("utf8").strip().split("\n")
+    exit_code = 0
+    for file in files:
+        file_perms = int(oct(os.stat(file).st_mode)[-3:])
+        if file_perms != ALLOWED_PERMISSION_EXECUTABLES:
+            print(
+                f"""File "{file}" contains a shebang line, but has the file permission {file_perms} instead of the expected executable permission {ALLOWED_PERMISSION_EXECUTABLES}. Do "chmod {ALLOWED_PERMISSION_EXECUTABLES} {file}" (or remove the shebang line)."""
```

Comment:
This requirement holds for shebang files that expected to be executed directly.

To avoid this requirement for files that are not expected to be excluded directly but still have shebangs consider adding exclusions for:
* `*.py` files that do not contain `if __name__ == '__main__':` or `if __name__ == "__main__":`: these are typically modules that are not expected to be executed directly.
* `*.bash` files: these are typically expected to be sourced and not executed directly.

It could be argued that shebangs should be removed from these types of files, but I think shebangs could still be useful to allow for language/dialect auto-detection in editors, `shellcheck`, etc.

Reply:
Updated. I wonder if some other file types should also be skipped along with `*bash`, I notice the test is failing for some files like `contrib/init/bitcoind.init`, `contrib/init/bitcoind.openrc`, `contrib/devtools/split-debug.sh.in"` and I'm thinking they might fit into the same category of files which are sourced but not executed?

There is also quite a few `.sh` files in `ci/test/` like `ci/test/00_setup_env_android.sh` which fail the test, but I'm not sure if those really should be forced to be executable or not?

Thanks very much for the initial review and feedback :)

Reply:
> Updated. I wonder if some other file types should also be skipped along with `*bash`? I notice the test is failing for some files like `contrib/init/bitcoind.init`, `contrib/init/bitcoind.openrc`, `contrib/devtools/split-debug.sh.in` and I'm thinking they might fit into the same category of files which are sourced but not executed?

Sounds reasonable. Skipping the requirement for `*.{init,openrc,sh.in}` should be fine.

> There is also quite a few `.sh` files in `ci/test/` like `ci/test/00_setup_env_android.sh` which fail the test, but I'm not sure if those really should be forced to be executable or not?

Perhaps we should simply make these executable to avoid the complexity of making the `+x` requirement depend also on directory (in addition to `#!` presence and suffix).

Reply:
Thanks! updated the condition to skip the requirement for `*.{init,openrc,sh.in}` files.

For the remaining [files failing the test](https://github.com/bitcoin/bitcoin/pull/21740/checks), what would be the correct way to go about updating those to comply with the test? Should it be a second commit on this PR?

Thanks again for the review and guidance.